src/logging/writers/ascii/Ascii.cc

```diff
@@ -16,6 +28,146 @@ using namespace threading;
 using threading::Value;
 using threading::Field;
 
+static constexpr auto shadow_file_prefix = ".shadow.";
+
+/**
+ * Information about an leftover log file: that is, one that a previous
+ * process was in the middle of writing, but never completed a rotation
+ * for whatever reason (prematurely crashed/killed).
+ */
+struct LeftoverLog {
+	/*
+	 * Name of leftover log, relative to working dir.
+	 */
+	std::string filename;
+
+	/*
+	 * File extension of the leftover log (e.g. ".log").
+	 */
+	std::string extension;
+
+	/*
+	 * Name of shadow file associated with the log.
+	 * The shadow file's existence is what indicates the presence of
+	 * an "leftover log" and may contain the name of a postprocessing
+	 * function that's supposed to be called after rotating (only
+	 * named if that function differs from the default).  Upon
+	 * completing a rotation, the shadow file can be deleted.
+	 */
+	std::string shadow_filename;
+
+	/**
+	 * Name of a function to call to postprocess the log file after
+	 * rotating.
+	 */
+	std::string post_proc_func;
+
+	/**
+	 * The time at which the shadow file was created.  This is used
+	 * as the log file's "opening time" for rotation purposes.
+	 */
+	time_t open_time;
+
+	/**
+	 * Time of the log file's last modification.  This is used
+	 * as the log file's "closing time" for rotation purposes.
+	 */
+	time_t close_time;
+
+	/**
+	 * Set the an error message explaining any error that happened while
+	 * trying to parse the shadow file and construct an object.
+	 */
+	std::string error;
+
+	/**
+	 * Return the name of the log without file extension (this is
+	 * what's called "path" in other logging framework parlance).
+	 * E.g. the name of "conn.log" is just "conn".
+	 */
+	std::string Name() const
```

Comment:
Any reason not to call this `Path` then?

Reply:
The reason was probably just that the "path" terminology of the logging framework doesn't match my expectations of it being related to real filesystem paths, but I'll change it to `Path()` for consistency.